modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sServerEndpoint.scala

```diff
@@ -121,26 +121,24 @@ private[smithy4s] class SmithyHttp4sServerEndpointImpl[F[_], Op[_, _, _, _, _],
       errorTransformation(other).flatMap(F.raiseError)
   }
 
-
-
-  // format: off
-  private val extractInput: (Metadata, Request[F]) ==> I = {
+  private val extractInput: (Metadata, Request[F]) => F[I] = {
     inputMetadataDecoder.total match {
       case Some(totalDecoder) =>
-        Kleisli(totalDecoder.decode(_: Metadata).liftTo[F]).local(_._1)
+        (metadata, request) =>
+          request.body.compile.drain *>
+            totalDecoder.decode(metadata).liftTo[F]
       case None =>
         // NB : only compiling the input codec if the data cannot be
         // totally extracted from the metadata.
-        implicit val inputCodec = entityCompiler.compilePartialEntityDecoder(inputSchema, entityCache)
-        Kleisli { case (metadata, request) =>
+        (metadata, request) =>
+          implicit val inputCodec =
+            entityCompiler.compilePartialEntityDecoder(inputSchema, entityCache)
```

Comment:
randomly scrolling, I think this still has the laziness issue (you only compile after the request). It's probably fine for perf now that we cache the compilers, but doing the compilation ahead of time seems more consistent with how we do it in other places

Reply:
resolved